data/src/main/java/io/gnosis/data/backend/dto/TransactionDto.kt

```diff
@@ -1,16 +1,60 @@
 package io.gnosis.data.backend.dto
 
+import pm.gnosis.model.Solidity
+
 data class DataDecodedDto(
     val method: String,
-    val parameters: List<ParamsDto>?
+    val parameters: List<ParamDto>?
 )
 
-data class ParamsDto(
-    val name: String,
-    val type: String,
-    val value: Any
+data class ValueDecodedDto(
+    val operation: Operation,
+    val to: Solidity.Address,
+    val value: Long,
+    val data: String,
+    val dataDecoded: DataDecodedDto?
 )
 
+sealed class ParamDto {
+    abstract val type: String
+    abstract val name: String
+    abstract val value: Any?
+
+    data class AddressParam(
+        override val type: String,
+        override val name: String,
+        override val value: Solidity.Address
+    ) : ParamDto()
+
+    data class ArrayParam(
+        override val type: String,
+        override val name: String,
+        override val value: List<Any>
+    ) : ParamDto()
+
+    data class BytesParam(
+        override val type: String,
+        override val name: String,
+        override val value: String,
+        val valueDecoded: List<ValueDecodedDto>?
```

Comment:
Value decoded ist not always a list ... It can be basically anything in this case it should fallback to null here. I would add a test for this

Reply:
Does valueDecoded have the structure of ValueDecodedDto? So it is should be at least a ValueDecodedDto object or null? What do you mean by saying it can be anything? 
`valueDecoded` is nullable, so it should fallback to null if that''s what you mean.

Reply:
If it is not a list of this object moshi will throw an error, you would need to catch that error in your custom adapter and set it to null.

And it can be literally anything ... E.g. `Any`

Reply:
To be fair this is not used in v1 of the data decoding epic afaik, so you could also just igbore it

Reply:
Done (setting valueDecoded to null if not a list). I will add tests

Reply:
Even if valueDecoded could be Any we would care only about specific cases on the client-side, otherwise, it would be hard to parse/display such data. Is this correct?

Reply:
In v1 we dont care about the decoded value, see https://github.com/gnosis/safe/issues/329

Reply:
sure, but what will we do for the next versions, where we will sho multisend data? Specify specific polymorphic subtypes?

Reply:
That is something you should worry about in the next version not now

Reply:
I would like to anticipate change. That's what good software is about

Reply:
but for now, everything is in place and #329 can be implemented

Reply:
>I would like to anticipate change. That's what good software is about

You are implementing logic for features that have not been fully speced out and have little impact on this feature. You could leave out the field and it would have 0 impact here. This is overthinking and slows down development. I do agree that there is a fine line and that you should check how future potential changes can impact current implementation (that is why I mentioned it should fallback to null in an error case). The rest is unknown ... as the issues have not been discussed in detail. Good software *development* is not only about thinking about edge cases, but also about *not overspecifying* (and knowing when not to do things). 

The client gateway returns it, because it is used in that form by the web client, but maybe we decide while discussing v2 of data decoding that we want to use a different structure and do some preprocessing on the client gateway (who knows).

Reply:
good point